src/apex/queuing.rs

```diff
@@ -77,48 +77,118 @@ pub mod abstraction {
+    {
+        fn as_ref(&self) -> &QueuingPortSender<S> {
+            &self.0
+        }
+    }
+
+    impl<const MSG_SIZE: MessageSize, const NB_MSGS: MessageRange, S: ApexQueuingPortP4Ext>
+        TryFrom<QueuingPortSender<S>> for ConstQueuingPortSender<MSG_SIZE, NB_MSGS, S>
+    {
+        type Error = Error;
+
+        fn try_from(port: QueuingPortSender<S>) -> Result<Self, Self::Error> {
+            if port.msg_size != MSG_SIZE || port.nb_msgs != NB_MSGS {
+                return Err(Error::InvalidConfig);
+            }
+
+            Ok(ConstQueuingPortSender(port))
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    pub struct ConstQueuingPortReceiver<
+        const MSG_SIZE: MessageSize,
+        const NB_MSGS: MessageRange,
+        Q: ApexQueuingPortP4Ext,
+    >(QueuingPortReceiver<Q>);
+
+    impl<const MSG_SIZE: MessageSize, const NB_MSGS: MessageRange, S: ApexQueuingPortP4Ext>
+        AsRef<QueuingPortReceiver<S>> for ConstQueuingPortReceiver<MSG_SIZE, NB_MSGS, S>
+    {
+        fn as_ref(&self) -> &QueuingPortReceiver<S> {
+            &self.0
+        }
+    }
+
+    impl<const MSG_SIZE: MessageSize, const NB_MSGS: MessageRange, S: ApexQueuingPortP4Ext>
+        TryFrom<QueuingPortReceiver<S>> for ConstQueuingPortReceiver<MSG_SIZE, NB_MSGS, S>
+    {
+        type Error = Error;
+
+        fn try_from(port: QueuingPortReceiver<S>) -> Result<Self, Self::Error> {
+            if port.msg_size != MSG_SIZE || port.nb_msgs != NB_MSGS {
+                return Err(Error::InvalidConfig);
+            }
+
+            Ok(ConstQueuingPortReceiver(port))
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct QueuingPortSender<Q: ApexQueuingPortP4Ext> {
         _b: PhantomData<AtomicPtr<Q>>,
         id: QueuingPortId,
+        msg_size: MessageSize,
+        nb_msgs: MessageRange,
     }
 
-    impl<const MSG_SIZE: MessageSize, const NB_MSGS: MessageRange, S: ApexQueuingPortP4Ext> Clone
-        for QueuingPortSender<MSG_SIZE, NB_MSGS, S>
-    {
+    impl<S: ApexQueuingPortP4Ext> Clone for QueuingPortSender<S> {
         fn clone(&self) -> Self {
             Self {
                 _b: self._b,
                 id: self.id,
+                msg_size: self.msg_size,
+                nb_msgs: self.nb_msgs,
             }
         }
     }
 
+    impl<const MSG_SIZE: MessageSize, const NB_MSGS: MessageRange, S: ApexQueuingPortP4Ext>
+        From<ConstQueuingPortSender<MSG_SIZE, NB_MSGS, S>> for QueuingPortSender<S>
+    {
+        fn from(port: ConstQueuingPortSender<MSG_SIZE, NB_MSGS, S>) -> Self {
+            port.0
+        }
+    }
+
     #[derive(Debug)]
-    pub struct QueuingPortReceiver<
-        const MSG_SIZE: MessageSize,
-        const NB_MSGS: MessageRange,
-        Q: ApexQueuingPortP4Ext,
-    > {
+    pub struct QueuingPortReceiver<Q: ApexQueuingPortP4Ext> {
         _b: PhantomData<AtomicPtr<Q>>,
         id: QueuingPortId,
+        msg_size: MessageSize,
+        nb_msgs: MessageRange,
```

Comment:
We did not require this before, and it is not used in any checks, just for the new function for querying the info about the port `range`. This increases the memory usage for storing queuing ports quite a bit. I am not sure, it is justified. When would a partition ever need this information and it is not sufficient to query the HV for it?

Reply:
Ok, I overlooked the conversion from the non-const variant. It also was there before as part of the const-generic. Also, I'm not sure if it could be "optimized away" or inlined anyway if the value / function is const.